File: 백준/Gold/14502. 연구소/연구소.py
'''
바이러스는 상하좌우빈칸으로 모두 퍼져나갈 수 있다 
새로 세울수있는 벽은 3개
바이러스가 퍼질 수 없는 안전영역이 최대가되는 값은?
2근처에 벽을 놓고 다 하나씩 돈다. ->최대값ㅇ르 구한다
1. 벽을 만든다.
2. 바이러스가 퍼진걸 체크한다. 
'''
import sys
from collections import deque
import copy
def make_wall(block): #벽만들기(DFS..)
    if block==3: #벽을 다 세웠다면
        bfs()
        return
    for i in range(n):
        for j in range(m):
            if mapp[i][j]==0:
                mapp[i][j]=1
                make_wall(block+1) #벽을 하나씩 세워보자
                mapp[i][j]=0 #그리고 다시 초기화
    

def bfs():
    queue=deque()
    test_mapp=copy.deepcopy(mapp)
    for i in range(n):
        for j in range(m):
            if test_mapp[i][j]==2:
                queue.append((i,j))

    while queue:
        x,y = queue.popleft()
        for i in range(4):
            nx=x+dx[i]
            ny=y+dy[i]
            if 0<=nx<n and 0<=ny<m and test_mapp[nx][ny]==0:
                test_mapp[nx][ny]=2
                queue.append((nx,ny))
    
    global result
    cnt=0
    for i in range(n):
        for k in range(m):
            if test_mapp[i][k] == 0:
                cnt +=1
    result=max(cnt,result)
    
n,m=map(int,sys.stdin.readline().split()) #세로, 가로
mapp=[list(map(int,sys.stdin.readline().split())) for _ in range(n)]
dx=[-1,0,1,0]
dy=[0,-1,0,1]
result=0
make_wall(0)
print(result)

File: 백준/Gold/14502. 연구소/연구소.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int mn=Integer.MIN_VALUE;
    static int n,m;
    static int[][] mapp;
    static int[] dx = new int[] {-1, 0, 1, 0};
    static int[] dy = new int[] {0, 1, 0,-1};


    public static void main(String[] args) throws IOException {
        //0의 크기 구하기
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken()); //세로
        m = Integer.parseInt(st.nextToken()); //가로

        mapp = new int[n][m];

        for (int i=0; i<n; i++) {
            StringTokenizer st1 = new StringTokenizer(br.readLine());
            for (int j=0; j<m; j++) {
                mapp[i][j]=Integer.parseInt(st1.nextToken());
            }
        }

        dfs(0);
        System.out.println(mn);
    }
    public static void dfs(int block) {
        if (block==3) {
            bfs();
            return;
        }

        for (int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                if (mapp[i][j]==0) {
                    mapp[i][j]=1;
                    dfs(block+1);
                    mapp[i][j]=0;
                }
            }
        }
    }


    private static void bfs() {
        //현재 맵에서 안전 얼만큼??연결??
        Queue<int[]> q = new LinkedList<>();

        for (int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                if (mapp[i][j]==2) {
                    q.add(new int[]{i,j});
                }
            }
        }

        //원본을 바꾸지 않기 위해 카피맵 사용
        int[][] copymap = new int[n][m];

        for (int i = 0; i < n; i++) {
            copymap[i] = mapp[i].clone();
        }

        while (!q.isEmpty()) {
            int[] temp =q.poll();
            int x= temp[0];
            int y = temp[1];

            for (int k=0; k<4; k++) {
                int nx = x+dx[k];
                int ny = y+dy[k];

                if (nx>=0 && nx<n && ny>=0 && ny<m ) {
                    if (copymap[nx][ny]==0) {
                        copymap[nx][ny]=2;
                        q.add(new int[] {nx,ny});
                    }
                }

            }

        }

        safe(copymap);
    }

    public static void safe(int[][] copymap) {
        int hap =0;
        for (int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                if (copymap[i][j]==0) {
                    hap++;
                }
            }
        }
        mn = Math.max(hap,mn);




    }


}
